packages/hint-doctype/src/hint.ts

```diff
@@ -0,0 +1,150 @@
+ */
+
+import { Category } from 'hint/dist/src/lib/enums/category';
+import { HintScope } from 'hint/dist/src/lib/enums/hintscope';
+import { HintContext } from 'hint/dist/src/lib/hint-context';
+import { IHint, HintMetadata, FetchEnd, ProblemLocation } from 'hint/dist/src/lib/types';
+import { debug as d } from 'hint/dist/src/lib/utils/debug';
+import { MatchInformation } from './types';
+
+const debug: debug.IDebugger = d(__filename);
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+export default class implements IHint {
+
+    public static readonly meta: HintMetadata = {
+        docs: {
+            category: Category.interoperability,
+            description: `This hint checks if the HTML is using the most modern DOCTYPE.`
+        },
+        id: 'doctype',
+        schema: [],
+        scope: HintScope.any
+    }
+
+    public constructor(context: HintContext) {
+        const regExpFactory = (isFlexible = false) => {
+            return new RegExp(`(<!doctype\\s+(html)\\s*(\\s+system\\s+"about:legacy-compat")?\\s*?>)${isFlexible ? '(.+)?' : ''}`, 'gi');
+        };
+
+        const correctLine = 0;
+        const doctypeRegExp = regExpFactory();
+        const doctypeRegExpFlexible = regExpFactory(true);
+
+        const defaultProblemLocation: ProblemLocation = {
+            column: 0,
+            line: 0
+        };
+
+        const report = async (resource: string, message: string, problemLocation = defaultProblemLocation): Promise<void> => {
+            await context.report(resource, null, message, undefined, problemLocation);
+        };
+
+        const getCurrentDoctypeProblemLocation = (text: string): ProblemLocation => {
+            const lines = text.split('\n');
+            const location = defaultProblemLocation;
+            let found = false;
+
+            lines.forEach((line: string, i: number): void => {
+                if (found) {
+                    return;
+                }
+
+                const matched = doctypeRegExpFlexible.exec(line);
+
+                if (matched){
+                    found = true;
+                    location.line = i;
+                    location.column = matched.index;
+                }
+            });
+
+            return location;
+        };
+
+        const getMatchInformation = (text: string): MatchInformation => {
+            return {
+                location: getCurrentDoctypeProblemLocation(text),
+                matches: text.match(doctypeRegExpFlexible)
+            };
+        };
+
+        const checkNoDoctypeInContent = async (matchInfo: MatchInformation, resource: string): Promise<boolean> => {
+            if (matchInfo.matches && matchInfo.matches.length > 0) {
+                return true;
+            }
+
+            await report(resource, 'The resource does not contain a valid DOCTYPE.');
```

Comment:
Nit: I think including an example would help someone fix the problem faster.
```suggestion
            await report(resource, 'The resource does not contain a valid DOCTYPE (e.g. `<!doctype html>`).');
```